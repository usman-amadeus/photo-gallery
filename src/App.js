import "react-photoswipe/lib/photoswipe.css";
import "./index.css";

import React, { useEffect, useState } from "react";
import axios from "axios";
import InfiniteScroll from "react-infinite-scroller";
import { PhotoSwipeGallery } from "react-photoswipe";

export const APIURL = "https://jsonplaceholder.typicode.com/photos";
export const fetchImages = (page = 1) => axios.get(`${APIURL}?albumId=${page}`);

let page = 0;

/**
 * Image gallery class
 */
function ImageGallery() {
  const [images, setImages] = useState([]);
  const [isInitialized, setIsInitialized] = useState(false);
  const [isImageViewerOpen, setIsImageViewerOpen] = useState(false);
  const [error, setError] = useState(false);

  /**
   * Get images from API
   */
  const getImages = async () => {
    page++;
    setError(false);
    try {
      const result = await fetchImages(page);
      setImages(images.concat(parseImages(result.data)));
    } catch (error) {
      setError(true);
    }
    setIsInitialized(true);
  };

  /**
   * Parse images to be able to be displayed by PhotoSwipeGallery
   *
   * @return list of parsed images with proper model
   */
  const parseImages = (data) => {
    return data.reduce((acc, image) => {
      acc.push({
        ...image,
        src: image.thumbnailUrl,
        thumbnail: image.thumbnailUrl,
        w: image.width ? image.width : 1200,
        h: image.height ? image.height : 900,
        title: image.id,
      });
      return acc;
    }, []);
  };

  /**
   * Close image viewer
   */
  const closeImageViewer = () => {
    setIsImageViewerOpen(false);
  };

  /**
   * Thumbnails content
   */
  const getThumbnailContent = (image) => (
    <img alt={image.title} src={image.thumbnail} width={180} height={150} />
  );

  useEffect(() => {
    if (!isInitialized) {
      getImages();
    }
  });

  return (
    <div className="homepage">
      <InfiniteScroll
        pageStart={page}
        loadMore={getImages}
        loader={<h4 style={{ textAlign: "center" }}>Loading...</h4>}
        hasMore={!!images.length}
        threshold={100}
      >
        {isInitialized ? (
          <PhotoSwipeGallery
            items={images}
            thumbnailContent={getThumbnailContent}
            isOpen={isImageViewerOpen}
            onClose={closeImageViewer}
          />
        ) : (
          <h4 style={{ textAlign: "center" }}>Loading...</h4>
        )}
      </InfiniteScroll>

      {error && (
        <h4 style={{ color: `red`, textAlign: "center" }}>
          Oops, error occured!
        </h4>
      )}
    </div>
  );
}
export default ImageGallery;
