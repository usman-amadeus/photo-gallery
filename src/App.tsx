import "react-photoswipe/lib/photoswipe.css";
import "./index.css";

import React, { useEffect, useState } from "react";
import axios from "axios";
import InfiniteScroll from "react-infinite-scroller";
import { PhotoSwipeGallery, PhotoSwipeGalleryItem } from "react-photoswipe";

export const APIURL = "https://jsonplaceholder.typicode.com/photos";
export const fetchImages = (page = 1) => axios.get(`${APIURL}?albumId=${page}`);
/**
 * Parse images to be able to be displayed by PhotoSwipeGallery
 *
 * @return list of parsed images with proper model
 */
export const parseImages = (data: ApiImage[]): PhotoSwipeGalleryItem[] => {
  return data.reduce((acc: PhotoSwipeGalleryItem[], image: ApiImage) => {
    acc.push({
      ...image,
      src: image.thumbnailUrl,
      thumbnail: image.thumbnailUrl,
      w: image.width ? image.width: 1200,
      h: image.height ? image.height : 900
    });
    return acc;
  }, []);
};


export interface ApiImage {
  albumId: number;
  id: number;
  title: string;
  url: string;
  thumbnailUrl: string;
  width?: number;
  height?: number;
}

let page = 0;

/**
 * Image gallery class
 */
function ImageGallery() {
  const [images, setImages] = useState([] as PhotoSwipeGalleryItem[]);
  const [isInitialized, setIsInitialized] = useState(false);
  const [isImageViewerOpen, setIsImageViewerOpen] = useState(false);
  const [error, setError] = useState(false);

  /**
   * Get images from API
   */
  const getImages = async () => {
    page++;
    setError(false);
    try {
      const result = await fetchImages(page);
      setImages([...images, ...parseImages(result.data)]);
    } catch (error) {
      setError(true);
    }
    setIsInitialized(true);
  };

  /**
   * Close image viewer
   */
  const closeImageViewer = () => {
    setIsImageViewerOpen(false);
  };

  /**
   * Thumbnails content
   */
  const getThumbnailContent = (image: PhotoSwipeGalleryItem) => (
    <img src={image.thumbnail} width={180} height={150} alt=''/>
  );

  useEffect(() => {
    if (!isInitialized) {
      getImages();
    }
  });

  return (
    <div className="homepage">
      <InfiniteScroll
        pageStart={page}
        loadMore={getImages}
        loader={<h4 style={{ textAlign: "center" }}>Loading...</h4>}
        hasMore={!!images.length}
        threshold={100}
      >
        {isInitialized ? (
          <PhotoSwipeGallery
            items={images}
            thumbnailContent={getThumbnailContent}
            isOpen={isImageViewerOpen}
            onClose={closeImageViewer}
          />
        ) : (
          <h4 style={{ textAlign: "center" }}>Loading...</h4>
        )}
      </InfiniteScroll>

      {error && (
        <h4 style={{ color: `red`, textAlign: "center" }}>
          Oops, error occured!
        </h4>
      )}
    </div>
  );
}
export default ImageGallery;
