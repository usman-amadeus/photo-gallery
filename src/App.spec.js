import React from "react";
import nock from "nock";
import TestRenderer from "react-test-renderer";
import ImageGallery from "./App";
import { fetchImages } from "./App";

describe("ImageGallery", () => {
  describe("Fetch images", () => {
    const data = [
      {
        id: "01",
        title: "Avengers Infinity War",
      },
      {
        id: "02",
        title: "Avengers Endgame",
      },
    ];

    it("should give proper title of the image", async () => {
      nock("http://localhost")
        .get("/photos")
        .query({ page: 1 })
        .reply(200, {
          result: { data },
        })
      const results = await fetchImages(1);
      expect(results.data[0].title).toEqual(
        "accusamus beatae ad facilis cum similique qui sunt"
      );
    });
  });

  describe("ImageGallery", () => {
    test("snapshot renders", () => {
        const component = TestRenderer.create(<ImageGallery />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
});
