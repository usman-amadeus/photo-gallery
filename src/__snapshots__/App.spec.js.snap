// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGallery ImageGallery snapshot renders 1`] = `
<div
  className="homepage"
>
  <div>
    <h4
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      Loading...
    </h4>
  </div>
</div>
`;

exports[`ImageGallery snapshot renders 1`] = `
<div
  className="HomePage"
>
  <div>
    <div>
      Â 
    </div>
  </div>
</div>
`;

exports[`Test api ImageGallery snapshot renders 1`] = `
<div
  className="homepage"
>
  <div>
    <h4
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      Loading...
    </h4>
  </div>
</div>
`;
